### PR TITLE
ICU-20494 Fixes to very large magnitude exponents in number parsing.

### PR DESCRIPTION
**Pull request by sffc**
##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20494
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [x] Tests included
- [ ] Documentation is changed or added

### CODE REVIEW

**Comment by sffc** [comment]
Ping

**Comment by sffc** [comment]
I rewrote the PR.  It fixes the Android behavior bug, plus also adds more related tests and fixes some other issues discovered while testing.

Notes:

- The change in `DecimalQuantity.applyMaxInteger` fixes an array/buffer overrun that I noticed while working on the other part of this bug.
- The new behavior for Infinity and NaN in ICU4C Formattable is consistent with ICU < 62.  There is not an equivalent thing to change in Java.
- The change in `DecimalQuantity.adjustMagnitude` ("Make sure that precision + scale won't overflow") was done in ICU4C in 33a0fa7 but was never done in ICU4J.

File: icu4j/main/classes/core/src/com/ibm/icu/impl/number/DecimalQuantity_DualStorageBCD.java [hunk 1]
```diff
@@ -252,17 +252,16 @@ protected BigDecimal bcdToBigDecimal() {
                 tempLong = tempLong * 10 + getDigitPos(shift);
             }
             BigDecimal result = BigDecimal.valueOf(tempLong);
-            try {
+            // Test that the new scale fits inside the BigDecimal
+            long newScale = result.scale() + scale;
+            if (newScale > Integer.MAX_VALUE || newScale <= Integer.MIN_VALUE) {
```
[the remark this answers is not in the document]
Comment (sffc) [reply icu4j/main/classes/core/src/com/ibm/icu/impl/number/DecimalQuantity_DualStorageBCD.java]: Yep.  I fixed this and a number of other issues.

**Comment by sffc** [comment]
Also @pedberg-icu as FYI.  I fixed the behavior of parsing a large exponent in ICU4C.  When you try to parse something like "1E99999999999999999", ICU < 62 behavior is to return "Infinity" from Formattable getDecimalNumber, but ICU 62-63 returned "0" instead.  This change makes ICU 64 return "Infinity" like ICU 61.

**Comment by pedberg-icu** [comment]
> Also @pedberg-icu as FYI. I fixed the behavior of parsing a large exponent in ICU4C. When you try to parse something like "1E99999999999999999", ICU < 62 behavior is to return "Infinity" from Formattable getDecimalNumber, but ICU 62-63 returned "0" instead. This change makes ICU 64 return "Infinity" like ICU 61.

Thanks! There is probably a bug in my incoming Apple ICU compatibility queue about that that I have  not got to analyzing yet... :-)

**Comment by sffc** [comment]
OK, I'm going to merge this based on Peter's review.  Victor/Andy, if you have any comments, let's fix them in a follow-up PR.

**Pull request merged**